File: src/codility/slice/MaxProfit.java
package codility.slice;

public class MaxProfit {

    public static void main(String[] args) {
        new MaxProfit().solution(new int[]{});
    }

    public int solution(int[] A) {
        int[] diff = new int[A.length];

        for (int i = 0;i <A.length-1;i++) {
            diff[i] = A[i +1] - A[i];
        }

        int maxProfit =0;

        int currentProfit = 0;
        for (int d : diff) {
            currentProfit += d;
            currentProfit = Math.max(currentProfit, 0);
            maxProfit = Math.max(maxProfit, currentProfit);
        }

        return maxProfit;

    }

}

File: src/codility/kotlin/MaxProfit.kt
package codility.kotlin

fun main() {

}

class MaxProfit {

    fun solution(A: IntArray): Int {
        // write your code in Kotlin

        if (A.isEmpty()) {
            return 0
        }

        var buy = A[0]
        var maxProfit = 0

        for ( i in 1 until A.size) {
            if (A[i] < buy) {
                buy = A[i]
            } else {
                maxProfit = Math.max(maxProfit, A[i] - buy)
            }

        }
        return maxProfit

    }
}